plugins/usbpro/DMXUSBDevice.cpp

```diff
@@ -0,0 +1,269 @@
+                                     OLA_UNUSED uint16_t device_id,
+                                     uint32_t serial,
+                                     uint16_t firmware_version,
+                                     unsigned int fps_limit):
+    UsbSerialDevice(owner, name, widget),
+    m_plugin_adaptor(plugin_adaptor),
+    m_dmxusb_widget(widget),
+    m_serial(),
+    m_got_parameters(false),
+    m_got_extended_parameters(false),
+    m_break_time(0),
+    m_mab_time(0),
+    m_rate(0),
+    m_out_ports(0),
+    m_in_ports(0),
+    m_fps_limit(fps_limit) {
+  m_str << std::setfill('0');
+  uint8_t *ptr = reinterpret_cast<uint8_t*>(&serial);
+  for (int i = UsbProWidgetInformation::SERIAL_LENGTH - 1; i >= 0; i--) {
+    int digit = (10 * (ptr[i] & 0xf0) >> 4) + (ptr[i] & 0x0f);
+    m_str <<  std::setw(2)  << digit;
+  }
+  m_serial = m_str.str();
+  m_str.str("");
+  m_str << "Serial #: " << m_serial << ", Firmware "
+      << (firmware_version >> 8) << "." << (firmware_version & 0xff);
+
+
+  m_dmxusb_widget->GetParameters(NewSingleCallback(
+    this,
+    &DMXUSBDevice::UpdateParams));
+  m_dmxusb_widget->GetExtendedParameters(NewSingleCallback(
+    this,
+    &DMXUSBDevice::UpdateExtendedParams));
+}
+
+
+/*
+ * Stop this device
+ */
+void DMXUSBDevice::PrePortStop() {
+  m_dmxusb_widget->Stop();
+}
+
+
+/*
+ * Handle device config messages
+ * @param controller An RpcController
+ * @param request the request data
+ * @param response the response to return
+ * @param done the closure to call once the request is complete
+ */
+void DMXUSBDevice::Configure(RpcController *controller,
+                                  const string &request,
+                                  string *response,
+                                  ConfigureCallback *done) {
+  Request request_pb;
+  if (!request_pb.ParseFromString(request)) {
+    controller->SetFailed("Invalid Request");
+    done->Run();
+    return;
+  }
+
+  switch (request_pb.type()) {
+    case ola::plugin::usbpro::Request::USBPRO_PARAMETER_REQUEST:
+      HandleParametersRequest(controller, &request_pb, response, done);
+      break;
+    case ola::plugin::usbpro::Request::USBPRO_SERIAL_REQUEST:
+      HandleSerialRequest(controller, &request_pb, response, done);
+      break;
+    default:
+      controller->SetFailed("Invalid Request");
+      done->Run();
+  }
+}
+
+
+/**
+ * Update the cached param values
+ */
+void DMXUSBDevice::UpdateParams(bool status,
+                                     const usb_pro_parameters &params) {
+  if (status) {
+    m_got_parameters = true;
+    m_break_time = params.break_time;
+    m_mab_time = params.mab_time;
+    m_rate = params.rate;
+  }
+}
+
+
+void DMXUSBDevice::UpdateExtendedParams(bool status,
+                                     const dmxusb_extended_parameters &params) {
+  if (status) {
+    m_got_extended_parameters = true;
+    m_out_ports = params.out_ports;
```

Comment:
Given your 47 limit in the tests there should be some sanity checks here or at port creation time.